File: backend/apps/common/document_folio.py
"""Folios de negocio Intrax: SERIE-número (COT / ODT / PRJ / POL)."""

from __future__ import annotations

import re

FOLIO_SERIE_COT = "COT"
FOLIO_SERIE_ODT = "ODT"
FOLIO_SERIE_PRJ = "PRJ"
FOLIO_SERIE_INS = "INS"
FOLIO_SERIE_POL = "POL"

_PREFIXED_FOLIO_RE = re.compile(r"^([A-Za-z]{3})-(\d+)$")


def format_document_folio(serie: str, idx, empty: str = "—") -> str:
    """Devuelve '{SERIE}-{n}' a partir del idx numérico."""
    if idx is None or idx == "":
        return empty
    try:
        if isinstance(idx, str):
            digits = re.sub(r"\D", "", idx)
            num = int(digits) if digits else 0
        else:
            num = int(idx)
    except (TypeError, ValueError):
        raw = str(idx).strip()
        return raw or empty
    if num <= 0:
        raw = str(idx).strip()
        return raw or empty
    s = (serie or "DOC").strip().upper() or "DOC"
    return f"{s}-{num}"


def resolve_document_folio(serie: str, folio, idx, empty: str = "—") -> str:
    """
    Si folio ya es SERIE-n, se respeta.
    Si folio es solo dígitos o está vacío, formatea con serie + idx/folio.
    Folios libres no numéricos (legado) se muestran tal cual.
    """
    existing = (str(folio).strip() if folio is not None else "")
    if existing:
        m = _PREFIXED_FOLIO_RE.match(existing)
        if m:
            return f"{m.group(1).upper()}-{m.group(2)}"
        if not existing.isdigit():
            return existing
        return format_document_folio(serie, existing, empty)
    return format_document_folio(serie, idx, empty)

File: frontend/src/utils/documentFolio.ts
/**
 * Folios de negocio Intrax: SERIE-número (COT / ODT / PRJ).
 * El número sigue siendo el `idx` interno; el prefijo es solo etiqueta visible.
 */

export const FOLIO_SERIE = {
  cotizacion: "COT",
  orden: "ODT",
  proyecto: "PRJ",
  poliza: "POL",
} as const;

export type DocumentFolioSerie = (typeof FOLIO_SERIE)[keyof typeof FOLIO_SERIE];

const PREFIXED_FOLIO_RE = /^([A-Za-z]{3})-(\d+)$/;

export function formatDocumentFolio(
  serie: DocumentFolioSerie | string,
  n: number | string | null | undefined,
  empty = "—"
): string {
  if (n == null || n === "") return empty;
  const num = typeof n === "number" ? n : Number(String(n).replace(/\D/g, ""));
  if (!Number.isFinite(num) || num <= 0) {
    const raw = String(n).trim();
    return raw || empty;
  }
  const s = String(serie || "").trim().toUpperCase() || "DOC";
  return `${s}-${Math.trunc(num)}`;
}

/** Si `folio` ya es SERIE-n, se respeta; si no, formatea con la serie dada y `idx`. */
export function resolveDocumentFolio(
  serie: DocumentFolioSerie | string,
  folio: string | null | undefined,
  idx: number | string | null | undefined,
  empty = "—"
): string {
  const existing = String(folio ?? "").trim();
  if (existing && PREFIXED_FOLIO_RE.test(existing)) {
    const m = existing.match(PREFIXED_FOLIO_RE)!;
    return `${m[1].toUpperCase()}-${m[2]}`;
  }
  if (existing && !/^\d+$/.test(existing)) {
    // Folio libre no numérico (legado): mostrar tal cual.
    return existing;
  }
  const fromExistingDigits = existing && /^\d+$/.test(existing) ? existing : idx;
  return formatDocumentFolio(serie, fromExistingDigits, empty);
}

export function parseDocumentFolio(input: string): { serie?: string; n?: number } {
  const raw = String(input || "").trim().replace(/^#/, "");
  if (!raw) return {};
  const m = raw.match(PREFIXED_FOLIO_RE);
  if (m) {
    return { serie: m[1].toUpperCase(), n: Number(m[2]) };
  }
  const digits = raw.replace(/\D/g, "");
  if (digits) return { n: Number(digits) };
  return {};
}

/** true si haystack (folio o idx) coincide con query flexible. */
export function matchesDocumentFolio(
  haystack: string | number | null | undefined,
  query: string
): boolean {
  const q = String(query || "").trim().toLowerCase();
  if (!q) return true;
  const h = String(haystack ?? "").trim().toLowerCase();
  if (!h) return false;
  if (h.includes(q) || h.replace(/^#/, "").includes(q.replace(/^#/, ""))) return true;

  const pq = parseDocumentFolio(q);
  const ph = parseDocumentFolio(h);
  if (pq.n != null && ph.n != null && pq.n === ph.n) {
    if (!pq.serie || !ph.serie) return true;
    return pq.serie === ph.serie;
  }
  if (pq.n != null && h === String(pq.n)) return true;
  return false;
}
